transform-plugins/src/main/java/co/cask/hydrator/plugin/ValueMapper.java

```diff
@@ -0,0 +1,228 @@
+import co.cask.cdap.api.plugin.PluginConfig;
+import co.cask.cdap.etl.api.Emitter;
+import co.cask.cdap.etl.api.LookupConfig;
+import co.cask.cdap.etl.api.LookupProvider;
+import co.cask.cdap.etl.api.Transform;
+import co.cask.cdap.etl.api.TransformContext;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonParser;
+import com.google.gson.JsonSyntaxException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Transforms records using custom mapping provided by the config.
+ */
+@Plugin(type = "transform")
+@Name("ValueMapper")
+@Description("Maps and convert record values from input sources using a mapping dataset")
+public class ValueMapper extends Transform<StructuredRecord, StructuredRecord> {
+
+  private static final Gson GSON = new GsonBuilder().create();
+  private static final Logger LOG = LoggerFactory.getLogger(ValueMapper.class);
+  private final Config config;
+  private final Map<Schema, Schema> schemaCache = Maps.newHashMap();
+  private final Map<String, String> fieldsMapping = Maps.newHashMap();
+  private final Map<String, ValueMapperLookUp> datasetMapping =
+          Maps.newHashMap();
+  private JsonArray mappingArray;
+
+  // for unit tests, otherwise config is injected by plugin framework.
+  public ValueMapper(Config config) {
+    this.config = config;
+  }
+
+  @Override
+  public void transform(StructuredRecord input, Emitter<StructuredRecord> emitter) throws Exception {
+    LOG.debug("in transform start :" + input.get("name"));
+    Schema outputSchema = getOutputSchema(input.getSchema());
+    StructuredRecord.Builder builder = createOutputRecords(input, outputSchema);
+    StructuredRecord ss = builder.build();
+    LOG.debug("in transform end :" + ss.get("designationName"));
+    emitter.emit(ss);
+  }
+
+  /**
+   * Creates output structure record using mapping dataset and set them to builder.
+   */
+  private StructuredRecord.Builder createOutputRecords(StructuredRecord input, Schema outputSchema) {
+
+    StructuredRecord.Builder builder = StructuredRecord.builder(outputSchema);
+    for (Schema.Field sourceField : input.getSchema().getFields()) {
+
+      LOG.debug("in createOutputRecords :" + input.get(sourceField.getName()
+      ).toString());
+      if (!input.get(sourceField.getName()).equals("NULL") &&
+              !input.get(sourceField.getName()).equals("EMPTY") &&
+              fieldsMapping.containsKey(sourceField.getName())) {
+        builder = lookUpAndSetValues(input, sourceField, builder);
+        LOG.debug("in createOutputRecords after :" + builder);
+      } else {
+        // for those source field whose values are either NULL or EMPTY
+        builder.set(sourceField.getName(), "" + input.get(sourceField.getName()
```

Comment:
don't convert to string, the field can be a non-string type.  Just use whatever the value is.